src/segment_reshape_plugin/plugin.py

```diff
@@ -61,40 +63,36 @@ def initGui(self) -> None:  # noqa N802 (qgis naming)
             message_log_name=tr("Segment reshape tool"),
         )
 
-        # Create toolbar
-        toolbar: QToolBar = iface.addToolBar(
+        toolbar = iface.addToolBar(
             self.tr("Segment reshape toolbar"),
         )
         toolbar.setObjectName("segment-reshape-toolbar")
 
-        # Add action
-        action = QAction(
+        self.segment_reshape_tool_action = QAction(
             QIcon(resources_path("icons/segment_reshape.svg")),
             self.tr("Reshape common segment"),
             iface.mainWindow(),
         )
-        action.setCheckable(True)
-        action.triggered.connect(self._segment_reshape_action_triggered)
 
-        self.segment_reshape_tool.setAction(action)
-        toolbar.addAction(action)
-        self.segment_reshape_tool_action = action
+        self.segment_reshape_tool_action.setCheckable(True)
+        self.segment_reshape_tool_action.triggered.connect(
+            self._activate_reshape_map_tool
+        )
+
+        self.segment_reshape_tool.setAction(self.segment_reshape_tool_action)
+        toolbar.addAction(self.segment_reshape_tool_action)
 
         self.toolbar = toolbar
 
     def unload(self) -> None:
         if self.toolbar is not None:
             self.toolbar.deleteLater()
-        del self.toolbar
         self.toolbar = None
 
         self._teardown_loggers()
         self._teardown_loggers = lambda: None
 
-    def _segment_reshape_action_triggered(self, checked: bool) -> None:
-        if checked is True:
-            if iface.mapCanvas().mapTool() != self.segment_reshape_tool:
-                iface.mapCanvas().setMapTool(self.segment_reshape_tool)
-            self.segment_reshape_tool.activate()
-        else:
-            self.segment_reshape_tool.deactivate()
```

Comment:
has this even worked previously, and/or do we still need this unchecked-state to deactivate the map tool?

Reply:
I tested this with the current release and it indeed didn't work as expected. The map tool was still active on canvas when it was tried to deactivate with the button. Only the button got unchecked.
I now cleaned the this further and implemented adding the map tool with `QgsAbstractMapToolHandler` and `iface.registerMapTool()`. This ensures that this map tool behaves like qgis native map tools which is that map tool can be deactivated only by selecting another map tool.
This now caused two plugin tests broken since the registermaptool method is not mocked in qgisinterface. Do you have suggestions how this should be handled?

Reply:
missing mocks on pytest-qgis `iface` are usually handled by adding the mock manually where necessary, see [picklayer fixture](https://github.com/nlsfi/pickLayer/blob/main/test/conftest.py#L39)

we havent really discussed whether to add these trivial missing apis to pytest-qgis, if they dont affect the global state of iface object?

 i.e. we probably dont need to assert here that it actually registered the map tool? compared to for example using `iface.activeLayer`, because this requires to hold a state on the mock iface object, and belongs more clearly to pytest-qgis code

Reply:
Yes, mocks of these simple methods are quite easy to add to the mocked iface. I also added the cadwidget here https://github.com/nlsfi/segment-reshape-qgis-plugin/blob/use-digitizing-map-tool/test/conftest.py#L38.
But this `registerMapToolHandler()` does quite a lot like connects the action to canvas.setmaptool etc. that are needed to test: https://github.com/nlsfi/segment-reshape-qgis-plugin/blob/93eb192143f56a3bbd730abbc3a79012e58f7227/test/plugin/test_plugin.py#L47

Reply:
in that case its probably fine to skip testing this exact behaviour, if it's an acceptable limitation of the unit tests where we dont have an actual iface object available. if we can trust the QGIS api works as intented and all this behaviour is provided by QGIS code, otherwise we would probably need to duplicate the c++ code behaviour on the mock?